### PR TITLE
feat(in-meeting-reactions): Reaction class

### PR DESCRIPTION
**Pull request by nataliadelmar**
# COMPLETES #2559 

[SPARK-393046](https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-393046)
This pull request addresses the creation of a reaction class that should be used meetings plugin to listen to reaction events inside a joined meeting. 

## by making the following changes:
- Opening a reactions class on meeting join and subscribing to listen to reaction events.
- Adding a subscribe and unsuscribe methods to reactions class (the later will be used once other PRs are merged, on meeting leave.
- Adding a process reaction method to reaction class to trigger a payload to the client about the received reaction.

### Change Type

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Tooling change
- [ ] Internal code refactor

## The following scenarios where tested

Unit and manual tests performed

### I certified that

- [X] I have read and followed [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request)
- [x] I discussed changes with code owners prior to submitting this pull request

- [x] I have not skipped any automated checks
- [ ] All existing and new tests passed
- [ ] I have updated the documentation accordingly

---

Make sure to have followed the [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request) before submitting.

### CODE REVIEW

File: packages/@webex/plugin-meetings/src/meeting/index.ts [hunk 1]
```diff
@@ -958,6 +960,15 @@ export default class Meeting extends StatelessWebexPlugin {
      */
     this.transcription = undefined;
 
+    /**
+     * Meeting transcription object
```
Comment (torgeadelin) [inline packages/@webex/plugin-meetings/src/meeting/index.ts:964]: Comment is wrong

File: packages/@webex/plugin-meetings/src/meeting/index.ts [hunk 2]
```diff
@@ -2060,7 +2071,8 @@ export default class Meeting extends StatelessWebexPlugin {
    * @returns {void}
    */
   handleDataChannelUrlChange(datachannelUrl) {
-    if (datachannelUrl && this.config.enableAutomaticLLM) {
+    // @ts-ignore - config coming from registerPlugin
```
Comment (torgeadelin) [inline packages/@webex/plugin-meetings/src/meeting/index.ts:2074]: why did you add this?

File: packages/@webex/plugin-meetings/src/meeting/index.ts [hunk 3]
```diff
@@ -3447,6 +3459,23 @@ export default class Meeting extends StatelessWebexPlugin {
     return false;
   }
 
+  /**
+ * Check if the meeting supports the Reactions
+ * @returns {boolean}
+ * @throws ReactionsNotSupportedError
```
Comment (torgeadelin) [inline packages/@webex/plugin-meetings/src/meeting/index.ts:3465]: I don't see it throwing anything? Maybe remove this comment

File: packages/@webex/plugin-meetings/src/meeting/index.ts [hunk 4]
```diff
@@ -3552,6 +3581,36 @@ export default class Meeting extends StatelessWebexPlugin {
     }
   }
 
+  /**
+   * Process reactions
+   * @private
+   * @returns {Promise<void>} a promise to process reactions
```
Comment (torgeadelin) [inline packages/@webex/plugin-meetings/src/meeting/index.ts:3587]: we are not returning anything in here, so put @reutns {void}

File: packages/@webex/plugin-meetings/src/meeting/index.ts [hunk 5]
```diff
@@ -3725,6 +3784,7 @@ export default class Meeting extends StatelessWebexPlugin {
         return join;
       })
       .then(async (join) => {
+        // @ts-ignore - config coming from registerPlugin
```
Comment (torgeadelin) [inline packages/@webex/plugin-meetings/src/meeting/index.ts:3787]: Why did you add this?

File: packages/@webex/plugin-meetings/src/meeting/index.ts [hunk 6]
```diff
@@ -3740,9 +3800,16 @@ export default class Meeting extends StatelessWebexPlugin {
               LoggerProxy.logger.info('Meeting:index#join --> enabled to recieve transcription!');
             }
           }
+          // @ts-ignore - config coming from registerPlugin
+          if (this.config.receiveReaction || options.receiveReaction) {
+            if (this.isReactionsSupported()) {
+              await this.receiveReaction();
```
Comment (torgeadelin) [inline packages/@webex/plugin-meetings/src/meeting/index.ts:3806]: I don't think we need to await. Also receiveReaction doesn't need to be async.
Also, maybe call this method receiveReaction`s`

File: packages/@webex/plugin-meetings/src/meeting/index.ts [hunk 7]
```diff
@@ -3740,9 +3800,16 @@ export default class Meeting extends StatelessWebexPlugin {
...
+              LoggerProxy.logger.info('Meeting:index#join --> enabled to recieve reaction!');
```
Comment (torgeadelin) [inline packages/@webex/plugin-meetings/src/meeting/index.ts:3807]: reactions

File: packages/@webex/plugin-meetings/src/reactions/constants.ts [hunk 8]
```diff
@@ -0,0 +1,3 @@
+export const REACTION_RELAY_TYPES = {
+    REACTION: 'react',
+  };
```
Comment (torgeadelin) [inline packages/@webex/plugin-meetings/src/reactions/constants.ts:3]: Format this file with prettier.

File: packages/@webex/plugin-meetings/src/reactions/index.ts [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+ * @class Reactions
+ */
+
+export default class Reactions {
+  members: Members;
+  llm: any;
+  processEvent: any;
+
+  /**
+   * @param {Members} members
```
Comment (torgeadelin) [inline packages/@webex/plugin-meetings/src/reactions/index.ts:17]: there are 2 parameters, add docs for llm too

File: packages/@webex/plugin-meetings/src/reactions/index.ts [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+   * @constructor
+   * @memberof Reactions
+   */
+  constructor(members: Members, llm: any) {
+    this.members = members;
+    this.llm = llm;
+  }
+
+  /**
+   * processes reaction and triggers event
```
Comment (torgeadelin) [inline packages/@webex/plugin-meetings/src/reactions/index.ts:27]: Capital P

File: packages/@webex/plugin-meetings/src/reactions/index.ts [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+    this.processEvent = (e: any) => {
+      if (e.data.relayType == REACTION_RELAY_TYPES.REACTION) {
+        const processedReaction = this.processReaction(e.data.reaction, e.data.sender);
+        callback(processedReaction);
+      }
+    };
+    this.llm.on('event:relay.event', this.processEvent);
+  }
+
+  unsubscribe() {
```
Comment (torgeadelin) [inline packages/@webex/plugin-meetings/src/reactions/index.ts:60]: Add docs for this method too

File: packages/@webex/plugin-meetings/test/unit/spec/meeting/index.js [hunk 12]
```diff
@@ -787,6 +787,36 @@ describe('plugin-meetings', () => {
           assert.calledOnce(meeting.transcription.closeSocket);
         });
       });
+      describe('#isReactionsSupported', () => {
+        it('should return false if the feature is not supported for the meeting', () => {
+          meeting.locusInfo.controls = {reactions: false};
```
Comment (torgeadelin) [inline packages/@webex/plugin-meetings/test/unit/spec/meeting/index.js:792]: technically the correct object is `{reactions: {enabled: false }}`

File: packages/@webex/plugin-meetings/test/unit/spec/meeting/index.js [hunk 13]
```diff
@@ -787,6 +787,36 @@ describe('plugin-meetings', () => {
...
+
+          assert.equal(meeting.isReactionsSupported(), false);
+        });
+        it('should return true if webex assitant is enabled', () => {
+          meeting.locusInfo.controls = {reactions: true};
```
Comment (torgeadelin) [inline packages/@webex/plugin-meetings/test/unit/spec/meeting/index.js:797]: technically the correct object is `{reactions: {enabled: false}}`

File: packages/@webex/plugin-meetings/test/unit/spec/reactions/index.ts [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+      });
+
+    describe('#processReaction', () => {
+        it('should stop listening on relay events', () => {
+          const result = reactions.processReaction(fakeReactionPayload, fakeSenderPayload);
+          assert.match(result, fakeProcessedReaction);
+        });
+      });
+  });
+});
```
Comment (torgeadelin) [inline packages/@webex/plugin-meetings/test/unit/spec/reactions/index.ts:126]: Also format this file with prettier :P

File: packages/webex/package.json [hunk 15]
```diff
@@ -46,6 +46,7 @@
     "@webex/common": "workspace:^",
     "@webex/internal-plugin-calendar": "workspace:^",
     "@webex/internal-plugin-device": "workspace:^",
+    "@webex/internal-plugin-llm": "workspace:^",
```
Comment (torgeadelin) [inline packages/webex/package.json:49]: these changes are a bit suss, they should already be in master. maybe try to merge `beta` into your branch

**Review by torgeadelin — CHANGES_REQUESTED** [review]
Looking gucci!
Just some small comments.

giphy

File: packages/@webex/plugin-meetings/src/meeting/index.ts (hunk 2 above)
Comment (nataliadelmar) [reply packages/@webex/plugin-meetings/src/meeting/index.ts:2074]: cause this.config was complaining; so I did what is done in this file about 20 times, like in line :2544

File: packages/@webex/plugin-meetings/src/meeting/index.ts (hunk 5 above)
Comment (nataliadelmar) [reply packages/@webex/plugin-meetings/src/meeting/index.ts:3787]: same as before, it is done this way for other this.config. This way, the file is not in red and you can focus in what is really typescript-wrong no?

File: packages/@webex/plugin-meetings/src/meeting/index.ts (hunk 6 above)
Comment (nataliadelmar) [reply packages/@webex/plugin-meetings/src/meeting/index.ts:3806]: oki, should I remove the async from the function declaration (line :3589)

File: packages/@webex/plugin-meetings/src/reactions/index.ts [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+
+/**
+ * @description Meeting Reactions feature.
+ * @exports
+ * @class Reactions
+ */
+
+export default class Reactions {
+  members: Members;
+  llm: any;
```
[the remark this answers is not in the document]
Comment (nataliadelmar) [reply packages/@webex/plugin-meetings/src/reactions/index.ts:13]: what do you mean? add what back? :)

File: packages/@webex/plugin-meetings/test/unit/spec/meeting/index.js (hunk 12 above)
Comment (nataliadelmar) [reply packages/@webex/plugin-meetings/test/unit/spec/meeting/index.js:792]: It has been working because we had in meetings/index.ts 
```
  isReactionsSupported() {
    if (this.locusInfo?.controls?.reactions) {
      return true;
    }
```

I changed it already to say
```
  isReactionsSupported() {
    if (this.locusInfo?.controls?.reactions.enabled) {
      return true;
    }
```

File: packages/@webex/plugin-meetings/src/meeting/index.ts (hunk 3 above)
Comment (maxinteger) [reply packages/@webex/plugin-meetings/src/meeting/index.ts:3465]: Looks like `isTranscriptionSupported` incorrect as well.

`@throws` needed only when the function throws error with `throw` keyword. Here the error handled with logging and returning `false`, no throwing

File: packages/@webex/plugin-meetings/src/reactions/index.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (maxinteger) [reply packages/@webex/plugin-meetings/src/reactions/index.ts:13]: try `LLMService` from llm plugins `import LLMService from '@webex/internal-plugin-llm';`

File: packages/@webex/plugin-meetings/src/reactions/index.ts [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+  };
+
+  /**
+   * Sends reactions data to given callback as it arrives.
+   *
+   * @param {Function} callback
+   * @returns {void}
+   */
+  subscribe(callback: Function) {
+    this.processEvent = (e: any) => {
```
Comment (maxinteger) [inline packages/@webex/plugin-meetings/src/reactions/index.ts:51]: the type of the `this.processEvent` already defined, no need for the `any`

```suggestion
    this.processEvent = (e) => {
```

File: packages/@webex/plugin-meetings/src/meeting/index.ts (hunk 5 above)
Comment (maxinteger) [reply packages/@webex/plugin-meetings/src/meeting/index.ts:3787]: This is trange. `this.config` should work, it is defined in `StatelessWebexPlugin` class but it does not. But `super` works: `super.config.enableAutomaticLLM` (at least no throw error).

I guess the main problem is, the super class is in a JS file.

The other strange thing is, we use named import `import StatelessWebexPlugin from '@webex/webex-core';` but `StatelessWebexPlugin` exported as default

**Review by maxinteger — CHANGES_REQUESTED** [review]

File: packages/@webex/plugin-meetings/src/meeting/index.ts (hunk 5 above)
Comment (nataliadelmar) [reply packages/@webex/plugin-meetings/src/meeting/index.ts:3787]: Mystery not solved from the sdk :(

File: packages/@webex/plugin-meetings/src/reactions/index.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (nataliadelmar) [reply packages/@webex/plugin-meetings/src/reactions/index.ts:13]: Module '"@webex/internal-plugin-llm/dist/llm.types"' has no exported member 'LLMService'.ts(2305) :(

File: packages/webex/package.json (hunk 15 above)
Comment (nataliadelmar) [reply packages/webex/package.json:49]: noup, it's still not on beta

File: packages/@webex/plugin-meetings/src/reactions/index.ts [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+/**
+ * @description Meeting Reactions feature.
+ * @exports
+ * @class Reactions
+ */
+
+export default class Reactions {
+  members: Members;
+  llm: any;
+  processEvent: any;
```
[the remark this answers is not in the document]
Comment (nataliadelmar) [reply packages/@webex/plugin-meetings/src/reactions/index.ts:14]: awesome!

File: packages/webex/package.json (hunk 15 above)
Comment (torgeadelin) [reply packages/webex/package.json:49]: i meant they should already be on `beta`

File: packages/@webex/plugin-meetings/src/reactions/index.ts [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+    return processedReaction;
+  };
+
+  /**
+   * Sends reactions data to given callback as it arrives.
+   *
+   * @param {Function} callback
+   * @returns {void}
+   */
+  subscribe(callback: Function) {
```
[the remark this answers is not in the document]
Comment (nataliadelmar) [reply packages/@webex/plugin-meetings/src/reactions/index.ts:51]: Do I need to change this in the documentation? If yes, how can I do that? It seems to be wrong to say `@param {(value: ProcessedReaction) => void} callback`

**Review by torgeadelin — APPROVED** [review]
LGTM!
Just make sure you address Laszlo's comments too.

File: packages/@webex/plugin-meetings/src/meeting/index.ts [hunk 20]
```diff
@@ -3964,9 +4012,21 @@ export default class Meeting extends StatelessWebexPlugin {
               LoggerProxy.logger.info('Meeting:index#join --> enabled to recieve transcription!');
             }
           }
+          // @ts-ignore - config coming from registerPlugin
+          if (this.config.receiveReactions || options.receiveReactions) {
+            if (this.isReactionsSupported()) {
```
Comment (maxinteger) [inline packages/@webex/plugin-meetings/src/meeting/index.ts:4017]: 1) these 2 `if` can be merged into 1
2) you should check the `this.reactions` value as well

```suggestion
          if (this.config.receiveReactions || options.receiveReactions && this.isReactionsSupported() && !this.reactions) {
```

if you have to create a new Reactions instance each time when `joinMeetingOptions` called, then you should unsubscribe the active reactions listener before you start a new one

File: packages/@webex/plugin-meetings/src/meeting/index.ts [hunk 21]
```diff
@@ -3964,9 +4012,21 @@ export default class Meeting extends StatelessWebexPlugin {
...
+              this.reactions = new Reactions(
```
Comment (maxinteger) [inline packages/@webex/plugin-meetings/src/meeting/index.ts:4018]: I can not find the place where the reactions object properly destroyed, for example when the user left the meeting.

At least it should unsubscribe the listener

File: packages/@webex/plugin-meetings/src/reactions/index.ts (hunk 19 above)
[the remark this answers is not in the document]
Comment (maxinteger) [reply packages/@webex/plugin-meetings/src/reactions/index.ts]: @nataliadelmar It looks verbose, but it is absolutely fine. But you can always create a new type and use that in the code/docs

**Review by maxinteger — CHANGES_REQUESTED** [review]
Only few comments about how the Reactions object cleanup handled

File: packages/@webex/plugin-meetings/src/reactions/index.ts (hunk 19 above)
[the remark this answers is not in the document]
Comment (nataliadelmar) [reply packages/@webex/plugin-meetings/src/reactions/index.ts]: PERFECT :) Added as a type

File: packages/@webex/plugin-meetings/src/meeting/index.ts (hunk 21 above)
Comment (nataliadelmar) [reply packages/@webex/plugin-meetings/src/meeting/index.ts:4018]: Catlin's PR is dealing with switching-off of the events listener and destroying the class on meeting leave :) 
https://github.com/webex/webex-js-sdk/pull/2579/files

File: packages/@webex/plugin-meetings/src/meeting/index.ts (hunk 20 above)
Comment (nataliadelmar) [reply packages/@webex/plugin-meetings/src/meeting/index.ts:4017]: sounds good; Ive separated the platform support conditions from the !this.reactions condition, so that, if the platforms doesn't support it, the corresponding error message is logged.

File: packages/@webex/plugin-meetings/test/unit/spec/reactions/index.ts [hunk 22]
```diff
@@ -0,0 +1,126 @@
...
+import uuid from 'uuid';
+import MockWebex from '@webex/test-helper-mock-webex';
+import LLMChannel from '@webex/internal-plugin-llm';
+import Mercury from '@webex/internal-plugin-mercury';
+import Meetings from '@webex/plugin-meetings/src/meetings';
+import Members from '@webex/plugin-meetings/src/members';
+import Reactions from '@webex/plugin-meetings/src/reactions';
+import {REACTION_RELAY_TYPES} from '@webex/plugin-meetings/src/reactions/constants';
+
+describe.only('plugin-meetings', () => {
```
Comment (shnaaz) [inline packages/@webex/plugin-meetings/test/unit/spec/reactions/index.ts:16]: should remove '.only'

**Review by arun3528 — CHANGES_REQUESTED** [review]
As llm has events coming from every where , I would suggest not save the members not the llm here ,  Just processing the date should be sufficient

better to have listen to llm in one place in meeting and have the llm push events based on the feature 

meeting(get event -> takes the member -> reaction class to process the data)

last time when i checked the SDK dom we were saving duplicate data in multiple places

File: packages/@webex/plugin-meetings/src/meeting/index.ts [hunk 23]
```diff
@@ -3964,9 +4012,24 @@ export default class Meeting extends StatelessWebexPlugin {
...
+          if (
+            // @ts-ignore - config coming from registerPlugin
+            (this.config.receiveReactions || options.receiveReactions) &&
+            this.isReactionsSupported()
+          ) {
+            if (!this.reactions) {
+              this.reactions = new Reactions(
+                this.members,
+                // @ts-ignore
+                this.webex.internal.llm
```
[the remark this answers is not in the document]
Comment (torgeadelin) [reply packages/@webex/plugin-meetings/src/meeting/index.ts]: we've refactored this

File: packages/@webex/plugin-meetings/src/meeting/index.ts [hunk 24]
```diff
@@ -3775,6 +3800,29 @@ export default class Meeting extends StatelessWebexPlugin {
...
   }
 
+  /**
+   * Process reactions
+   * @private
+   * @returns {void} a promise to process reactions
+   */
+  private receiveReactions() {
+    try {
+      this.reactions.subscribe((payload) => {
```
[the remark this answers is not in the document]
Comment (torgeadelin) [reply packages/@webex/plugin-meetings/src/meeting/index.ts]: we've refactored this

File: packages/@webex/plugin-meetings/src/meeting/index.ts [hunk 25]
```diff
@@ -3775,6 +3790,44 @@ export default class Meeting extends StatelessWebexPlugin {
...
+   * @returns {void}
+   */
+  private processRelayEvent = (e: any): void => {
+    switch (e.data.relayType) {
+      case REACTION_RELAY_TYPES.REACTION:
+        if (
+          // @ts-ignore - config coming from registerPlugin
+          (this.config.receiveReactions || options.receiveReactions) &&
+          this.isReactionsSupported()
+        ) {
```
Comment (maxinteger) [inline packages/@webex/plugin-meetings/src/meeting/index.ts:3806]: I guess these values can not change during the meeting. 

Please register the `processRelayEvent` listener only when it is needed

```ts
        if (
          // @ts-ignore - config coming from registerPlugin
          (this.config.receiveReactions || options.receiveReactions) &&
          this.isReactionsSupported()
        ) {
           this.webex.internal.llm.on('event:relay.event', this.processRelayEvent);
        }
```

Note: you do not need the same condition for the `off` because it is OK to remove unregistered listener

File: packages/@webex/plugin-meetings/src/meeting/index.ts [hunk 26]
```diff
@@ -3775,6 +3790,44 @@ export default class Meeting extends StatelessWebexPlugin {
     }
   }
 
+  /**
+   * Callback called when a relay event is received from meeting LLM Connection
+   * @param {any} e Event object coming from LLM Connection
+   * @private
+   * @returns {void}
+   */
+  private processRelayEvent = (e: any): void => {
```
Comment (maxinteger) [inline packages/@webex/plugin-meetings/src/meeting/index.ts:3799]: We can do better than `any` here

In the `reactions.types.ts`:

```ts

export type RelayEvent = {
  data: RelayEventData
};

type RelayEventData = {
  relayType: typeof REACTION_RELAY_TYPES['REACTION']
  reaction: Reaction;
  sender: Sender;
}

```

The good thing with this pattern, you can extend it later like this:

```ts

type RelayEventData =
  | {
      relayType: typeof REACTION_RELAY_TYPES['REACTION'];
      reaction: Reaction;
      sender: Sender;
    }
  | {
      relayType: typeof REACTION_RELAY_TYPES['TYPE_2'];
      something: boolean;
    };
```
TS understands the link between these type variants. It called Discriminated Unions, mode details here:
https://www.typescriptlang.org/docs/handbook/2/narrowing.html#discriminated-unions

**Review by maxinteger — CHANGES_REQUESTED** [review]
Nice simplification! I have only 2 minor comments

File: packages/@webex/plugin-meetings/src/meeting/index.ts (hunk 26 above)
Comment (srinijay98) [reply packages/@webex/plugin-meetings/src/meeting/index.ts:3799]: wow, thats a good read, thanks @maxinteger

File: packages/@webex/plugin-meetings/src/meeting/index.ts (hunk 21 above)
Comment (torgeadelin) [reply packages/@webex/plugin-meetings/src/meeting/index.ts]: no longer needed because of the refactor

File: packages/@webex/plugin-meetings/src/meeting/index.ts (hunk 26 above)
Comment (torgeadelin) [reply packages/@webex/plugin-meetings/src/meeting/index.ts:3799]: thanks for the suggestion!
Addressed

File: packages/@webex/plugin-meetings/src/meeting/index.ts (hunk 25 above)
Comment (torgeadelin) [reply packages/@webex/plugin-meetings/src/meeting/index.ts:3812]: Discussed 1:1
image


TLDR; = leave as is.

**Review by maxinteger — APPROVED** [review]

**Review by arun3528 — APPROVED** [review]
Looks good so there is no need for reaction class for now we wil create it if we need in future
